New article "Overview of WebView2 features and APIs" (#1971)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/concepts/overview-features-apis.md

##### [C#](#tab/c-sharp)

* [CoreWebView2Controller.CoreWebView2 Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2controller.corewebview2)
* [CoreWebView2Controller.Close Method](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2controller.close)
* [CoreWebView2Environment.CreateCoreWebView2ControllerAsync Method](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment.createcorewebview2controllerasync)<!--2 overloads-->

##### [C++](#tab/cpp)

* [ICoreWebView2Controller::CoreWebView2 property (get)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#get_corewebview2)<!--no put-->
* [ICoreWebView2Controller::Close method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#close)
* [ICoreWebView2Environment::CreateCoreWebView2Controller method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment#createcorewebview2controller)

---

<!-- ====================================================================== -->

> ### Window management

Author: ```suggestion
## Window management
```
Revert to h2?  We only have used h2s, and (logical) h3s (tagged as h4's for distinct smaller css sizing), in this file.  The recent demotion from h2 to h3 makes this whole section a child of the h2 "Rendering WebView2 in non-framework apps", which is probably wrong.
@nishitha-burman

   * [ICoreWebView2AcceleratorKeyPressedEventArgs interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2acceleratorkeypressedeventargs)

---

<!-- ------------------------------ -->
#### Default background color

WebView2 can specify a default background color.  This can be any opaque color or transparent.  This color will be used if the HTML page doesn't set its own background color.

##### [C#](#tab/c-sharp)

* [CoreWebView2Controller.DefaultBackgroundColor Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2controller.defaultbackgroundcolor)

##### [C++](#tab/cpp)

* [ICoreWebView2Controller2::DefaultBackgroundColor property (get](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller2#get_defaultbackgroundcolor), [put)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller2#put_defaultbackgroundcolor)

---

<!-- ====================================================================== -->

> ### Composition-based rendering

Author: ```suggestion
## Composition-based rendering
```
```suggestion
## Window management
```
Revert to h2?  We only have used h2s, and (logical) h3s (tagged as h4's for distinct smaller css sizing), in this file.  The recent demotion from h2 to h3 makes this whole section a child of the h2 "Rendering WebView2 in non-framework apps", which is probably wrong.
@nishitha-burman

Author: Resolved: re-promoted to h2 as "Rendering WebView2 using Composition"

##### [C++](#tab/cpp)

* [ICoreWebView2Environment7::UserDataFolder property (get)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment7#get_userdatafolder)<!--no put-->
* [ICoreWebView2EnvironmentOptions2::ExclusiveUserDataFolderAccess property (get](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions2#get_exclusiveuserdatafolderaccess), [put)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions2#put_exclusiveuserdatafolderaccess)
* [ICoreWebView2Profile2::ClearBrowsingData method](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2profile2?view=webview2-1.0.1245.22#clearbrowsingdata)
* [ICoreWebView2Profile2::ClearBrowsingDataAll method](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2profile2?view=webview2-1.0.1245.22#clearbrowsingdataall)
* [ICoreWebView2Profile2::ClearBrowsingDataInTimeRange method](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2profile2?view=webview2-1.0.1245.22#clearbrowsingdataintimerange)

> * [ICoreWebView2Environment10 interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment10)<!--TODO: why is this link here, which members?-->

Author: @nishitha-burman  - why link to ICoreWebView2Environment10?  similar q for C# above

Reviewer: @mikehoffms this interface is used for multiple user profile support.

Author: Finding, part A: C#: It would|would not be feasible to link directly to each of the _ members, which are: (tbd).
Finding, part B: C++: It would be possible to link directly to each of the 3 members.

Author: implemented per above.
https://review.docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/overview-features-apis?branch=pr-en-us-1971&tabs=win32cpp

> ---

Reviewer 2: Do we have a systematic way for folks to add to this doc as new apis come out? Might be worthwhile to add a step like this to the api addition one note page.

Reviewer: Yeah we have discussed adding to this doc each month during the release so whoever does the release notes also adds to this doc if needed.

title: Overview of WebView2 features and APIs
description: Overview of WebView2 features and APIs.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
ms.date: 06/30/2022
---
# Overview of WebView2 features and APIs

> By including WebView2 in your app, your app can use various methods and properties provided by the WebView2 control to easily add many kinds of browser features.

Reviewer 2: A little confused with the intro to this doc. I would hope for a little more guidance into why I would want to read this doc. In addition why are we only speaking about browser features in this intro, when there are many other things you can do with wv2? Try something like below.

```suggestion
Embedding the WebView2 control in your app gives it access to various methods and properties provided in WebView2 interfaces. We have hundreds of APIs that provide a vast set of capabilities ranging from enhance native capabilities or allow you to modify browser experiences. The goal of this document is to provide a high-level grouping for WebView2 apis to help contextualize the different things you can do with WebView2. 
```

Author: @jasonstephen15 - "Guidance on why I would want to read this doc" - Here is previous draft text from top of article, maybe we should add something along these lines, but tighter.

This article provides:
*  A high-level understanding of the capabilities of the WebView2 technology.
*  Concepts about what you can do with the WebView2 APIs.
*  High-level description of WebView2 technology for app developers.
*  Better fundamentals and understanding of WebView2 technology.

Author: Embedding the WebView2 control in your app gives it access to various methods and properties provided via the WebView2 interfaces.  WebView2 has hundreds of APIs that provide a vast set of capabilities, ranging from enhancing your app's native-platform capabilities to enabling your app to modify browser experiences.  This article provides a high-level grouping of WebView2 APIs to help you understand the different things you can do using WebView2.

-->

The `CoreWebView2Environment`, `CoreWebView2Controller`, and `CoreWebView2` classes (or equivalent interfaces) work together so your app can host a WebView2 browser control and access its browser features.  These three large classes expose a wide range of APIs that your host app can access to provide many categories of browser-related features for your users.

*  The `CoreWebView2Environment` class represents a group of WebView2 controls that share the same WebView2 browser process, user data folder, and renderer.  From this `CoreWebView2Environment` class, you create pairs of `CoreWebView2Controller` and `CoreWebView2` instances.
*  The `CoreWebView2Controller` class is responsible for hosting-related functionality such as window focus, visibility, size, and input, where your app hosts the WebView2 control.
*  The `CoreWebView2` class is for the web-specific parts of the WebView2 control, including networking, navigation, script, and parsing and rendering HTML.

<!-- / keep sync'd -->

See also:
* [Main classes for WebView2: Environment, Controller, and Core](environment-controller-core.md)

> ##### [C#](#tab/c-sharp)

Reviewer 2: We have C#, and C++ tabs in this table, but we usually split our apis up based on the api set/runtime? Like win32/.net/winrt. Like maybe there is a place where c++ development can happen in C#, or C++ can be used alongside winrt apis, in which case developers may get confused? I actually don't know enough to advise here, maybe @david-risney has thoughts?

Author: Please see how many sets of API Refs we want to link to in the present article's tab-sets:
https://docs.microsoft.com/en-us/microsoft-edge/webview2/webview2-api-reference#languages-and-frameworks

Author: Resolved: we're trying moving toward this model
https://docs.microsoft.com/en-us/microsoft-edge/webview2/release-notes?tabs=dotnetcsharp#promotions-1
The initial tent. commit, in middle tab, might have a single WinRT placeholder link to top of its API Ref, pending member links.

---

<!-- ------------------------------ -->
#### Parent window

WebView2 can be reparented to a different parent window handle (`HWND`).  WebView2 also needs to be notified when the app's position on the screen has changed.

##### [C#](#tab/c-sharp)

* [CoreWebview2Controller.NotifyParentWindowPositionChanged Method](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2controller.notifyparentwindowpositionchanged)
   * [CoreWebview2Controller.ParentWindow Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2controller.parentwindow)

##### [C++](#tab/cpp)

* [ICoreWebview2Controller::NotifyParentWindowPositionChanged method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#notifyparentwindowpositionchanged)
   * [ICoreWebview2Controller::ParentWindow property (get](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#get_parentwindow), [put)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#put_parentwindow)

---

<!-- ------------------------------ -->

> #### Keyboard accelerators

Reviewer 2: Why is this under window management?

Author: resolved: removed intermediate heading "Window mgmt", so this is now going to be directly under 
"Rendering WebView2 in non-framework apps"

   * [CoreWebView2AcceleratorKeyPressedEventArgs Class](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2acceleratorkeypressedeventargs)

##### [C++](#tab/cpp)

* [ICoreWebView2Settings3::AreBrowserAcceleratorKeysEnabled property (get](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2settings3#get_arebrowseracceleratorkeysenabled), [put)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2settings3#put_arebrowseracceleratorkeysenabled)
* [ICoreWebView2Controller::AcceleratorKeyPressed event (add](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#add_acceleratorkeypressed), [remove)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2controller#remove_acceleratorkeypressed)
   * [ICoreWebView2AcceleratorKeyPressedEventArgs interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2acceleratorkeypressedeventargs)

---

<!-- ------------------------------ -->

> #### Default background color

Reviewer 2: Also feel like background color should not go under window management

Author: Window management heading level has been dissolved 
so this is now directly under h2 "Rendering WebView2 in non-framework apps"

* [ICoreWebView2::OpenDevToolsWindow method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2#opendevtoolswindow)

Call:
* [ICoreWebView2::CallDevToolsProtocolMethod method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2#calldevtoolsprotocolmethod)
* [ICoreWebView2_11::CallDevToolsProtocolMethodForSession method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2_11#calldevtoolspotocolmethodforsession)

Receiver:
* [ICoreWebView2::GetDevToolsProtocolEventReceiver method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2#getdevtoolsprotocoleventreceiver)
   * [ICoreWebView2DevToolsProtocolEventReceiver interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2devtoolsprotocoleventreceiver)
   * [ICoreWebView2DevToolsProtocolEventReceivedEventArgs interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2devtoolsprotocoleventreceivedeventargs)

---

<!-- ====================================================================== -->
## See also

> 

Author: link to https://docs.microsoft.com/en-us/microsoft-edge/webview2/webview2-api-reference

* [ICoreWebView2Environment4::GetAutomationProviderForWindow method](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment4#getautomationproviderforwindow)<!--not in c#-->

---

<!-- ====================================================================== -->
## User data

Manage the user data folder (UDF), which is a folder stored on the user's machine, that contains data related to the host app and WebView2.  WebView2 apps use user data folders to store browser data, such as cookies, permissions, and cached resources.

See also:
* [Manage user data folders](user-data-folder.md)

##### [C#](#tab/c-sharp)

* [CoreWebView2Environment.UserDataFolder Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment.userdatafolder)
* [CoreWebView2EnvironmentOptions.ExclusiveUserDataFolderAccess Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environmentoptions.exclusiveuserdatafolderaccess)<!--new-->
* [CoreWebView2Profile Class](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2profile)<!--new-->

>    * `ClearBrowsingDataAsync` Methods

Author: list the 3 overloads here & c++

* [CoreWebView2Environment Class](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment)<!--new-->

>    * [CoreWebView2Environment.UserDataFolder Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment.userdatafolder)

Author: dup of above.  clean up this list

* [CoreWebView2Environment interface](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment)<!--new--><!--TODO: why is this link here, which members?-->

##### [C++](#tab/cpp)

* [ICoreWebView2Environment7::UserDataFolder property (get)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment7#get_userdatafolder)<!--no put-->
* [ICoreWebView2EnvironmentOptions2::ExclusiveUserDataFolderAccess property (get](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions2#get_exclusiveuserdatafolderaccess), [put)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions2#put_exclusiveuserdatafolderaccess)<!--new-->
* [ICoreWebView2Profile2 interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2profile2)<!--new-->

>    * `ClearBrowsingData` methods

Author: link to the 3 methods, for consistent look

The `NavigationStarting` event allows the app to cancel navigations in WebView2, including for frames.

##### [C#](#tab/c-sharp)

* [CoreWebView2.NavigationStarting Event](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2.navigationstarting)
   * [CoreWebView2NavigationStartingEventArgs Class](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2navigationstartingeventargs)
* [CoreWebView2.FrameNavigationStarting Event](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2.framenavigationstarting)
* [CoreWebView2Frame.NavigationStarting Event](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2frame.navigationstarting)
   * [CoreWebView2NavigationStartingEventArgs Class](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2navigationstartingeventargs)

##### [C++](#tab/cpp)

* [ICoreWebView2::NavigationStarting event (add](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2#add_navigationstarting), [remove)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2#remove_navigationstarting)

>    * [ICoreWebView2NavigationStartingEventArgs interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2navigationstartingeventargs)<!--TODO: used by above? yes according to Ref-->

Author: Args, or Args2?  same question for Frame at end of list

Reviewer 3: that's how our COM API does versioning. They get version numbers at the end to indicate newer versions. There's still one object that implements all the different version interfaces. So even though the event handler provides the v1 version of the interface, the end developer can get the v2 version of the interface from that same object. Probably no need to provide all the different versions of the same interface. Can probably link to just the latest or just the earliest interface

   * [CoreWebView2Environment.UserDataFolder Property](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment.userdatafolder)
* [CoreWebView2Environment interface](https://docs.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment)<!--new--><!--TODO: why is this link here, which members?-->

##### [C++](#tab/cpp)

* [ICoreWebView2Environment7::UserDataFolder property (get)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment7#get_userdatafolder)<!--no put-->
* [ICoreWebView2EnvironmentOptions2::ExclusiveUserDataFolderAccess property (get](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions2#get_exclusiveuserdatafolderaccess), [put)](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environmentoptions2#put_exclusiveuserdatafolderaccess)<!--new-->
* [ICoreWebView2Profile2 interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2profile2)<!--new-->
   * `ClearBrowsingData` methods

> * [ICoreWebView2Environment10 interface](https://docs.microsoft.com/microsoft-edge/webview2/reference/win32/icorewebview2environment10)<!--new--><!--TODO: why is this link here, which members?-->

Author: @nishitha-burman  - why link to ICoreWebView2Environment10?  similar q for C# above

Author: closing "outdated" version of this cmt

---
title: Overview of WebView2 features and APIs
description: Overview of WebView2 features and APIs.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: webview
ms.date: 07/07/2022
---
# Overview of WebView2 features and APIs

<!-- todo: re-sync table sentences -->

> Embedding the WebView2 control in your app gives your app access to various methods and properties that are provided through the WebView2 classes or interfaces.  WebView2 has hundreds of APIs that provide a vast set of capabilities, ranging from enhancing your app's native-platform capabilities,<!--such as [concrete/specific]--> to enabling your app to modify browser experiences.<!--reword "modify browser experiences"; ... such as [concrete/specific]-->  This article provides a high-level grouping of the WebView2 APIs to help you understand the different things you can do using WebView2.

Reviewer 4: ```suggestion
WebView2 contains hundreds of APIs that provide a vast set of capabilities, ranging from enhancing your app's native-platform capabilities ,<!--such as [concrete/specific]--> to enabling your app to modify browser experiences.<!--reword "modify browser experiences"; ... such as [concrete/specific]-->, or to enabling your app to modify browser experiences. Leveraging this technology gives your app access to various methods and properties that are provided through the WebView2 classes or interfaces.

This article provides a high-level grouping of the WebView2 APIs to help you understand the different things you can do using WebView2.
```

* [ICoreWebView2BasicAuthenticationResponse interface](/microsoft-edge/webview2/reference/win32/icorewebview2basicauthenticationresponse)
   * [ICoreWebView2HttpResponseHeaders interface](/microsoft-edge/webview2/reference/win32/icorewebview2httpresponseheaders)
* [ICoreWebView2HttpHeadersCollectionIterator interface](/microsoft-edge/webview2/reference/win32/icorewebview2httpheaderscollectioniterator)

---

<!-- ====================================================================== -->
## Rendering WebView2 in non-framework apps

Use these APIs to set up the WebView2 rendering system if your host app doesn't use a UI framework.  This rendering setup controls how WebView2 renders output into your host app, and how WebView2 handles input, focus, and accessibility.

<!-- TODO: make a simple flat list of the 3-4 scenarios, clearly differentiated/identified/described.  switch to table layout? -->
*  If you're using a UI framework for your app, you should use the WebView2 element that's provided by that UI framework, rather than using these APIs.

> *  If you're not using a UI framework for your app (for example, if you're using pure Win32 directly), or if your UI framework doesn't have a WebView2 element, then you need to create `CoreWebView2Controller` and render it into your app, using these APIS.

Reviewer: Nit: APIS --> APIs
